plugins/targetlocker/inmemory/inmemory.go

```diff
@@ -95,30 +111,48 @@ func broker(lockRequests, unlockRequests, checkLocksRequests <-chan request, don
 			}
 			req.err <- lockErr
 		case req := <-unlockRequests:
+			if err := validateRequest(req); err != nil {
+				req.err <- fmt.Errorf("unlock request: %w", err)
+			}
 			log.Debugf("Requested to transactionally unlock %d targets: %v", len(req.targets), req.targets)
+			var unlockErr error
 			for _, t := range req.targets {
-				if _, ok := locks[*t]; ok {
-					delete(locks, *t)
+				if l, ok := locks[*t]; ok {
+					if l.owner == req.owner {
+						delete(locks, *t)
+					} else {
+						unlockErr = fmt.Errorf("unlock request: denying unlock request from job ID %d on lock owned by job ID %d", req.owner, l.owner)
+					}
 				} else {
-					log.Debugf("Target is not locked, but received unlock request: %+v", t)
+					// XXX should we roll back the unlock if a target fails?
+					unlockErr = fmt.Errorf("unlock request: target is not locked: %+v", t)
+					break
 				}
 			}
-			req.err <- nil
+			req.err <- unlockErr
 		case req := <-checkLocksRequests:
-			log.Debugf("Requested to check locks for %d targets: %v", len(req.targets), req.targets)
+			if err := validateRequest(req); err != nil {
+				req.err <- fmt.Errorf("checklocks request: %w", err)
+			}
+			log.Debugf("Requested to check locks for %d targets by job ID %d: %v", len(req.targets), req.owner, req.targets)
 			locked := make([]*target.Target, 0)
 			notLocked := make([]*target.Target, 0)
 			for _, t := range req.targets {
 				if l, ok := locks[*t]; ok {
-					now := time.Now()
-					if now.After(l.expiresAt) {
-						// target was locked but lock expired, purge the entry
-						log.Debugf("Purged expired lock for target %+v. Lock time is %s, expiration timeout is %s", t, l.lockedAt, req.timeout)
-						delete(locks, *t)
-						notLocked = append(notLocked, t)
+					if l.owner == req.owner {
```

Comment:
checklocks also didn't check for the lock owner - now it returns correctly the locks that are owned by the request owner. All the rest is returned as non-locked or not-locked-by-you